### ISSUE TITLE
Proxy error when trying to enter new tags

### ISSUE THREAD

**Issue by kereid** [aodn/aatams#263]
Steps to reproduce:

try to enter a new tag

What should happen:

Tag is successfully entered

What does happen:

Get an error stating

The proxy server received an invalid response from an upstream server. The proxy server could handle the request. Reason is error reading from remote server

**Comment by jkburges** [comment]
Noting that I *suspect* that whatever's causing #262 is also causing this...  as well as not being able to access the detections screen.

Running the query `SELECT relation::regclass, * FROM pg_locks WHERE NOT GRANTED;` against the DB while both of these problems were occurring showed many queries/processes waiting for locks to be granted.

The issue seems to have somehow resolved itself, running the above query returns 0 results and one can now access the detections screen.

**Comment by jkburges** [comment]
*Possibly* some correlation with the harvester which was running at the time.  I don't have any evidence for this apart from the fact that it was running!

**Comment by jkburges** [comment]
Whatever is causing this bug, it's pretty serious.  Basically, the system is mostly unusable whilst it's happening (i.e. the app cannot read from the DB).  Same applies to #262.

**Comment by julian1** [comment]
As discussed in #262 a harvester was running at the time of the reported issues. The harvester copies data using a join on the receiver materialized view. 

```
SELECT detection.id, receiver_deployment_id, timestamp, transmitter_id, detection.receiver_name  FROM aatams.detection LEFT JOIN aatams.receiver ON detection.receiver_name::text = receiver.receiver_name LEFT JOIN aatams.deployment_and_recovery deployment_and_recovery ON receiver.id = deployment_and_recovery.receiver_id AND detection.timestamp >= deployment_and_recovery.initialisationdatetime_timestamp AND detection.timestamp <= deployment_and_recovery.recoverydatetime_timestamp WHERE aatams.invalid_reason(detection.*, receiver.*, deployment_and_recovery.*) IS NULL
```

This action acquires a shared SharedLock on receiver.

```
aatams3  | | aatams.receiver                                          | aatams_acoustic | 00:00:00 | 00:17:10  | relation |       |            |               | 29584 | AccessShareLock     | t
```

When the user enters a new tag, a corresponding insert on the device table is created.

```
insert into device (version, comment, model_id, serial_number, code_map_id, expected_life_time_days, project_id, status, class, id) values ($2, $2, $3, $4, $5, $6, $7, $8, 'au.org.emii.aatams.Tag', $9)
```

A trigger on the device table is setup to update the receiver materialized view. 
```
\d aatams.device
...
Triggers:
    device_trigger AFTER INSERT OR DELETE OR UPDATE ON aatams.device FOR EACH STATEMENT EXECUTE PROCEDURE aatams.refresh_receiver_mv()

aatams3=# \df+ aatams.refresh_receiver_mv ;
                                                                                         List of functions
 Schema |        Name         | Result data type | Argument data types |  Type   | Security | Volatility | Owner  | Language |                      Source code                       | Description
--------+---------------------+------------------+---------------------+---------+----------+------------+--------+----------+--------------------------------------------------------+-------------
 aatams | refresh_receiver_mv | trigger          |                     | trigger | invoker  | volatile   | aatams | plpgsql  |                                                       +|
        |                     |                  |                     |         |          |            |        |          |                  BEGIN                                +|
        |                     |                  |                     |         |          |            |        |          |                    REFRESH MATERIALIZED VIEW receiver;+|
        |                     |                  |                     |         |          |            |        |          |                    RETURN NEW;                        +|
        |                     |                  |                     |         |          |            |        |          |                  END;                                 +|
        |                     |                  |                     |         |          |            |        |          |                                                        |
(1 row)

```

The materialized view update requires an exclusive lock on the table,

```
 aatams3  | | aatams.receiver                                          | aatams          | 00:01:44 | 00:02:06  | relation |       |            |               | 30332 | AccessExclusiveLock | f
```

According to postgres documentation http://www.postgresql.org/docs/9.3/static/explicit-locking.html, an AccessExclusiveLock conflicts with locks of all other modes. The device insert query therefore has to wait until the harvester has completed - perhaps many hours later. As the query hangs the browser UI times out, with a proxy error presumably from apache.

**Comment by jkburges** [comment]
> The harvester copies data using a join on the receiver materialized view

Do you think that if the harvester was changed to read from the `receiver view` (as opposed to *materialised* view), then this would fix this problem?

**Comment by julian1** [comment]
#### Tested fix

Postgres 9.4 includes supports for refreshing a materialized view concurrently. It works by computing new data in a background table, and generating a digest of insert/deletes actions against the materialized table - removing the very aggressive locking needs in 9.3.

To confirm the cause of the bad behavior, the test-instance was upgraded from 9.3 to 9.4 using pg_dump and pg_restore and the trigger function changed to use the new concurrency feature.
```
-- alter the trigger update function
create or replace function aatams.refresh_receiver_mv() returns trigger as $$
BEGIN
  refresh materialized view concurrently receiver;
 RETURN NEW;
END;
$$
language plpgsql volatile;

-- create needed index
create unique index on aatams.receiver(id);
```

Testing with a full-copy of production data showed that new tags could be created via the aatams UI at the same time the harvester was running. Monitoring of db queries/ locks also showed no collisions.

#### Suggested solutions,

1. Don't fix, but simply avoid running the aatams_acoustic_reporting harvester during times that the aatams application is known to be used.

2. Upgrade to postgres 9.4, and change the mat-view trigger function in liquibase to be concurrent.

3. Remove the receiver mat-view, and investigate ways to improve search performance. @jkburges discussed the possibility of splitting out the Tag and Receiver data into two tables, which may give a smaller (and therefore more performant) receiver table.

**Comment by julian1** [comment]
> Do you think that if the harvester was changed to read from the `receiver view` (as opposed to *materialised* view), then this would fix this problem?  

Reserving the mat-view for search queries only? Yes that makes sense to me. 

Although, I was just playing with the receiver view select. It's actually very fast,

```
  SELECT device.id,
               (((device_model.model_name)::text || '-'::text) || (device.serial_number)::text) AS receiver_name,
               device.organisation_id
               FROM (device          
               JOIN device_model ON ((device.model_id = device_model.id)))
               WHERE ((device.class)::text = 'au.org.emii.aatams.Receiver'::text);
```

All records retrieved in 8ms. 

Are we certain it needs to be a mat-view?

**Comment by jkburges** [comment]
> Are we certain it needs to be a mat-view?

No.  If you want to investigate changing it, you would need to make sure the tests at https://github.com/aodn/aatams/blob/master/doc/detections_performance_analysis/derived_validity/benchmark-initial.sql aren't badly affected (using same postgresql settings as on prod)

**Comment by julian1** [comment]
Some performance benchmarks, with the view changed, 
https://docs.google.com/spreadsheets/d/1SWHubkssPHxGFMKNktyvCBC0wFhBCiImcWAIpp22wDY

**Comment by julian1** [comment]
I tested that new tags can be entered with the harvester running using the non-materialized receivier view. It works correctly.

Note, on the nectar node I am testing with (4cpu, 16Mb), with postgres memory the same as prod, the app responds very slowly when the harvester is running.  For example, logging in took about 30seconds. The new tag action is about the same.

The code is on the branch, https://github.com/aodn/aatams/tree/remove_receiver_matview

**Comment by julian1** [comment]
So, another solution,

4. Remove the receiver matview using the migration, https://github.com/aodn/aatams/compare/remove_receiver_matview

I probably like this option the most, so long as search performance is judged to be adequate but @jkburges should choose.

**Comment by jkburges** [comment]
> so long as search performance is judged to be adequate

Looking at those performance benchmarks, generally the performance is around the same, but there are a couple of queries which seem to be badly affected.  I'm not sure that I want to make a call on whether or not performance is adequate without further discussion.

I would be a fan of removing the receiver matview also though, if we can get away with it.

@julian1, I'm not sure of the relevance of the second set of results (offset 0)?

**Comment by jkburges** [comment]
@xhoenner, what schedule does the harvester currently run on? I think a sensible thing to do for now would be to make sure that it runs only outside of business hours, if possible, to mitigate this problem.  I'm not saying that this is the final solution, just temporary (for the next couple of years :smile: )

**Comment by xhoenner** [comment]
@jkburges agreed, you'll have to ask @danfruehauf as I don't have access to that info (I think it's in chef-private)

**Comment by danfruehauf** [comment]
@xhoenner Which harvester?

By default all harvesters are queuing to run at 18:00.

**Comment by xhoenner** [comment]
> @xhoenner Which harvester?

aatams_acoustic_reporting

**Comment by danfruehauf** [comment]
```
    "cron": {
        "weekday": "5"
    },
    "rc": {
        "cron": {
            "day": "1,7,13,20"
        }
    }
```

Which means Friday in production, and the 1st, 7th, 13th and 20th in RC. That will queue on 18:00. I cannot guarantee the exact time it will start running though.

**Comment by jkburges** [comment]
In summary, a [before and after comparison of the change](https://docs.google.com/spreadsheets/d/117fNDAa9LjrQiVsAJI34mMMt1-e0oZkJnvQO2ypJYGE/edit#gid=0) https://github.com/aodn/aatams/compare/remove_receiver_matview show that generally, most queries slow down a little, a few slow down a lot and one quite significantly. 

The proposed solution is this:

* add monitoring for actual queries that the web-app is making to the DB, logging those that take over a certain threshold (say 20s)
* keep an eye on this for a few weeks
* after we have a decent record of real-world usage, deploy the change above, and then keep monitoring
* add a warning to the detections list view that some queries may take several minutes
* ensure that long request/responses will not time out due to proxies etc

**Comment by dnahodil** [comment]
Google doc with the results from testing we did (trying to reproduce @julian1's results): https://docs.google.com/spreadsheets/d/117fNDAa9LjrQiVsAJI34mMMt1-e0oZkJnvQO2ypJYGE/edit?usp=sharing

The script we used to clear the disk caches and run the tests:
``` bash
for i in {1..3};
  do sudo service postgresql stop && sync && echo 3 | sudo tee /proc/sys/vm/drop_caches && sudo service postgresql start && sudo -u postgres psql aatams3 < /home/dnahodil/benchmark-initial.sql > "run$i.txt";
done
```

**Comment by dnahodil** [comment]
The plan currently is to go down the route of upgrading Postgres to 9.4 and if there are still problems, look at making database changes.

@jkburges also investigated some other options regarding changing some joins to use new matviews instead of views.

**Comment by jkburges** [comment]
Fixed by upgrading to PostgreSQL 9.4
